File: snippets/go/usecases/microservices/bookingservice.go
package main

import (
	restate "github.com/restatedev/sdk-go"
)

type BookingRequest struct {
	HotelID  string `json:"hotel_id"`
	FlightID string `json:"flight_id"`
}

type BookingResult struct {
	Success bool `json:"success"`
}

func CancelHotel(hotelID string) error {
	// Simulate hotel cancellation
	return nil
}

func BookHotel(request BookingRequest) (restate.Void, error) {
	// Simulate hotel booking
	return restate.Void{}, nil
}

func CancelFlight(flightID string) error {
	// Simulate flight cancellation
	return nil
}

func BookFlight(request BookingRequest) (restate.Void, error) {
	// Simulate flight booking
	return restate.Void{}, nil
}

// <start_here>
type BookingService struct{}

func (BookingService) Reserve(ctx restate.Context, request BookingRequest) (res BookingResult, err error) {
	var compensations []func() error
	defer func() {
		if err != nil {
			for i := len(compensations) - 1; i >= 0; i-- {
				_, compErr := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
					return restate.Void{}, compensations[i]()
				})
				if compErr != nil {
					err = compErr
				}
			}
		}
	}()

	// Reserve hotel
	compensations = append(compensations, func() error { return CancelHotel(request.HotelID) })
	_, err = restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		return BookHotel(request)
	})
	if err != nil {
		return BookingResult{}, err
	}

	// Reserve flight
	compensations = append(compensations, func() error { return CancelFlight(request.FlightID) })
	_, err = restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		return BookFlight(request)
	})
	if err != nil {
		return BookingResult{}, err
	}

	return BookingResult{Success: true}, nil
}

// <end_here>

File: snippets/java/src/main/java/usecases/microservices/BookingService.java
package usecases.microservices;

import dev.restate.sdk.Restate;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import usecases.microservices.utils.BookingRequest;
import usecases.microservices.utils.BookingResult;

@Service
public class BookingService {

  // <start_here>
  @Handler
  public BookingResult reserve(BookingRequest request) {
    List<Runnable> compensations = new ArrayList<>();

    try {
      // Reserve hotel
      compensations.add(() -> cancelHotel(request.hotelId));
      Restate.run("book-hotel", () -> bookHotel(request));

      // Reserve flight
      compensations.add(() -> cancelFlight(request.flightId));
      Restate.run("book-flight", () -> bookFlight(request));

      return new BookingResult(true);
    } catch (Exception error) {
      // Run compensations in reverse order
      Collections.reverse(compensations);
      for (Runnable compensation : compensations) {
        Restate.run("compensation", compensation::run);
      }
      throw error;
    }
  }

  // <end_here>

  private void cancelHotel(String hotelId) {
    // Simulate hotel cancellation
  }

  private void bookHotel(BookingRequest request) {
    // Simulate hotel booking
  }

  private void cancelFlight(String flightId) {
    // Simulate flight cancellation
  }

  private void bookFlight(BookingRequest request) {
    // Simulate flight booking
  }
}
